nroff/nroffdoc/superman.1
.Id $Id: superman.1,v 1.1 2000-10-21 14:31:38 jon Exp $
.Im mantools utility
.TH superman 1
.XI superman - "Nroff/Troff file concatenation utility"
.SH NAME
superman \- Nroff/Troff file concatenation utility
.SH SYNOPSIS
.sp 1
.na
\fBsuperman\fR
\fC[\fB\-?\fC|\fB\-h\fC]\fR
\fC[\fB\-E\fR\ \fI<file>\fC]\fR
\fC[\fB\-e\fR\ \fI<file>\fC]\fR
\fC[\fB\-I\fC]\fR
\fC[\fB\-o\fR\ \fI<file>\fC]\fR
\fC[\fB\-q\fC]\fR
\fC[\fB\-v\fR\ \fI<level>\fC]\fR
\fInroff\ files\ \.\.\.\fR
.ad
.SH DESCRIPTION
\fBsuperman\fR  is a  Nroff/Troff  utility  used to  concatenate  source files
together  into a single file ready for  processing  by the other Nroff  tools.
The utility takes as input a list of Nroff files and generates a single output
file which contains all of the input files in a format described by
.Ht sm 4 .
.PP
The process of  constructing  a "Super Man" page, which  comprises a number of
individual manual pages enables the processing of the information by the other
tools to be significantly  increased as there is not a continual  accessing of
the disk to locate the files. During the build  process, a single  manual page
will be accessed  multiple  times by  different  utilities  (e.g. symbol table
building, library  construction and the conversion  process itself). The super
manual  page also makes it  extremely  easy to move a modules  pages round the
system during the build process as a single file large only needs to be moved,
rather than a lot of small individual files.
.PP
The utility  guarantees  to place all of the include files at the start of the
file,  provided that they have a \fB.tni\fR  extension to identify  them as an
Nroff include file.
.PP
The \fBsuperman\fR  utility also recognises other superman files, and performs
the necessary file  extraction  from other superman files when  constructing a
larger superman file.
.PP
The superman  process is a two pass  operation, the first pass collects all of
the include files  (\fB.tni\fR) and adds them to the superman file. The second
pass collects the main pages and adds them to the superman file.
.PP
The command options are defined as follows:-
.\" -?/-h help option
.so opt_h.tni
.\" -e -E : Error options.
.so opt_e.tni
.\" -I : Information.
.so opt_i.tni
.IP "\fB-o\fR \fI<file>\fR"
Specifies the name of the output file  \fI<file>\fR, if the output file is not
specified then the data is written to \fIstdout\fR.
.\" -q : Quiet mode.
.so opt_q.tni
.IP "\fB-v\fR \fI<level>\fR"
Specify the  verbosity  level of the utility.  \fI<level>\fR  is the verbosity
level which is specified  as a single digit and ranges from 0 through 9, where
0 is least  verbose.  Typically  level 1 is used to monitor the files that are
appended to the archive.
.SH EXAMPLE
The following example concatenates a group of Nroff manual pages into a single
file called \fCf.sm\fR:-
.CS
superman -o f.sm option.tni f1.1 f2.1 f3.1
.CE
Multiple superman files may be concatinated together as follows:-
.CS
superman -o foobar.sm foo.sm bar.sm
.CE
Alternatively, a composite  command line may be specified with different types
of source data, the utility will still operate correctly and generate a single
output file \fCcomposite.sm\fR, note that verbose is enabled:-
.CS
superman -v 1 -o composite.sm foo.sm bar.sm option.tni f1.1 f2.1 *.4
.CE 0
.SH NOTES
The Nroff tools described by this package suite all recognise superman files
.Ht sm 4 .
The UNIX  N/Troff  utilities  do not  understand  the file  format,  hence the
individual files should be extracted prior to submission to
.Hr groff 1
using a simple
.Hr nawk 1
script or program.
.SH BUGS
The utility does not recognise that a file has been  submitted  multiple times
if it is  submitted  via  another  superman  file,  hence the output  file may
contain  duplicate  copies of the same file. It is  considered to be the users
responsibility  to ensure that the input file stream is  submitted  correctly,
without duplicates.
.SH "SEE ALSO"
.na
.Hr groff 1
.Hr nawk 1
.Ht nr2html 1 ,
.Ht nr2rtf 1 ,
.Ht nrinfo 1 ,
.Hr nroff 1 ,
.Ht nrsearch 1 ,
.Ht sm 4 .
.ad
.FH


nroff/nroff/superman.c
/****************************************************************************
 *
 *			Copyright 1996-2004 Jon Green.
 *                         All Rights Reserved
 *
 *
 *  System        :
 *  Module        :
 *  Object Name   : $RCSfile: superman.c,v $
 *  Revision      : $Revision: 1.4 $
 *  Date          : $Date: 2004-02-07 19:29:49 $
 *  Author        : $Author: jon $
 *  Last Modified : <260202.1522>
 *
 *  Description
 *
 *  Notes
 *
 *  History
 *
 * 1.0.4h - JG 07/02/04 - Ported to HP-UX 11.0.
 * 1.0.4h - JG 03/01/04 - Ported to Sun Solaris 9.
 * 1.0.4g - JG 21/10/00 - Corrected the file extension comparison.
 * 1.0.4f - JG 20/04/97 - Added base name option.
 * 1.0.4e - JG 29/08/96 - Added UNIX file name utilities.
 * 1.0.4d - JG 05/01/95 - Added external references
 * 1.0.4c - JG 16/11/95 - Ported to UNIX world
 *
 ****************************************************************************
 *
 *  Copyright (c) 1996-2004 Jon Green.
 *
 *  All Rights Reserved.
 *
 * This  Document  may  not, in  whole  or in  part, be  copied,  photocopied,
 * reproduced,  translated,  or  reduced to any  electronic  medium or machine
 * readable form without prior written consent from Jon Green.
 *
 ****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#if ((defined _HPUX) || (defined _LINUX) || (defined _SUNOS))
#include <unistd.h>
#else
#include <getopt.h>
#endif

#include <utils.h>

/*
 */

#define MODULE_VERSION  "1.0.4i"
#define MODULE_NAME     "superman"

static char *progname = MODULE_NAME;    /* Name of the module */
static char *outFile = NULL;
static int  useBase = 0;                /* Use the base name */

static FILE *fo;                        /* Output File Pointer */
#ifdef _UNIX
static char *eolStr = "\n";
static char eofStr [2] = {0, 0};
#else
static char *eolStr = "\r\n";
static char eofStr [2] = {0x1A, 0};
#endif

#if 1
static int
readLine (FILE *fp, char *cbuf)
{
    int     c;
    int     len;

    if (fp == NULL)
        return (-1);

    cbuf [0] = '\0';
    len = 0;
    for (;;)
    {
        c = fgetc (fp);
        if ((c == EOF) || (c == 0x1A))
            return (-1);

        if (c == '\n') {
            cbuf [len] = '\0';
            return (len);
        }
        else if (c == '\r')
            continue;
        cbuf [len++] = c;
    }   /* End of 'while' */
}   /* End of 'readLine' */

static void
doSuperMan (int pass, char *fname)
{
    FILE    *fp;
    char    buf [1024];
    int     mode = 0;
    char    *smfname;
    int     lineNo;

    if ((fp = fopen (fname, "rb")) == NULL) {
        uError ("Cannot open file [%s] for reading.\n", fname);
        return;
    }

    uFileSet (&lineNo, fname);
    while (readLine (fp, buf) >= 0)
    {
        lineNo++;
        if (strncmp (".SUPERMANINC",buf, 12) == 0) {
            if (mode != 0)
                uError("Unexpected %s\n", buf);
            else if (pass == 1) {
                smfname = &buf [13];
                uInteractive ("Archiving [%s]", smfname);
                mode = 0x81;
            }
        }
        else if (strncmp (".SUPERMANFILE", buf, 13) == 0) {
            if (mode != 0)
                uError("Unexpected %s\n", buf);
            else if (pass == 2) {
                mode = 0x82;
                smfname = &buf [14];
                uInteractive ("Archiving [%s]", smfname);
            }
        }
        else if (((mode & 0x80) != 0) &&
                 (strncmp (".SUPERMANEND", buf, 12) == 0))
            mode |= 0x04;

        if ((mode & 0x80) != 0)
            fprintf (fo, "%s%s", buf, eolStr);
        if ((mode & 0x04) != 0)
            mode = 0;
    }
    if (mode != 0)
        uError ("Unexpected EOF in superman file\n");
    fclose (fp);
    uFileSet (NULL, NULL);
}

#endif

static void
processFile (int pass, char *filename)
{
    FILE *fp;
    int  i;
    int  isIncludeFile;
    int  isCopyMode;
    int  isCleanLine = 1;
    char *fname;
    char *ext;                          /* Extension component */
    char *base;                         /* Base name component */
    char *bname;                        /* The base name */

    uVerbose (1+pass, "Processing File.\n");

    fname = unifyFilename (filename);
    if (splitFilename (fname, NULL, NULL, &base, &ext) != 0)
        uFatal ("Cannot split filename [%s]\n", fname);
    if (useBase)
        bname = makeFilename (NULL, NULL, base, ext);
    else
        bname = fname;

    isIncludeFile = ((strcmp (ext, "so") == 0) || (strcmp (ext, "tni") == 0));

#if 1
    if ((isIncludeFile == 0) && (strcmp (ext, "sm") == 0))
    {
        doSuperMan (pass, filename);
        pass = -1;
    }
#endif

    if ((pass == 1) && isIncludeFile)
        isCopyMode = 1;
    else if ((pass == 2) && (isIncludeFile == 0))
        isCopyMode = 2;
    else
        isCopyMode = 0;

    if (isCopyMode)
    {
        /* Open the file first - make sure that it exists. */
        if ((fp = fopen (fname, "rb")) == NULL)
            uError ("Cannot open file [%s] for reading.\n", fname);
        else
        {
            int lineNo = 0;

            uFileSet (&lineNo, fname);
            if ((i = fgetc (fp)) == EOF)
                uWarn ("Zero Length File [%s] - skipping\n", fname);
            else
            {
                if (isCopyMode == 1) {
                    uInteractive ("Archiving [%s]", fname);
                    fprintf (fo, ".SUPERMANINC %s%s", bname, eolStr);
                }
                else {
                    uInteractive ("Archiving [%s]", fname);
                    fprintf (fo, ".SUPERMANFILE %s%s", bname, eolStr);
                }

                do {
                    if (i == 0x1A)
                        break;
                    if (i == '\n') {
                        fprintf (fo, "%s", eolStr);
                        isCleanLine = 1;
                        lineNo++;
                    }
                    else if (i != '\r') {
                        fputc (i, fo);
                        isCleanLine = 0;
                    }
                } while ((i = fgetc (fp)) != EOF);

                if (isCleanLine == 0)
                    fprintf (fo, "%s", eolStr);
                fprintf (fo, ".SUPERMANEND %s%s", bname, eolStr);
            }
            fclose (fp);
            uFileSet (NULL, NULL);
        }
    }
}

static void
usage (void)
{
    fprintf (stdout, "%s : Usage %s [Options] <files>\n", progname, progname);
    fprintf (stdout, "\n");
    fprintf (stdout, "-?/h      : Help - this page\n");
    fprintf (stdout, "-b        : Use basename or insertion\n");
    fprintf (stdout, "-E <file> : Log errors to <file> (append)\n");
    fprintf (stdout, "-e <file> : Log errors to <file> (re-write)\n");
    fprintf (stdout, "-I        : Version Information\n");
    fprintf (stdout, "-q        : Quiet mode\n");
    fprintf (stdout, "-o <file> : Output file name.\n");
    fprintf (stdout, "-v <level>: Verbose level.\n");
    exit (1);
}

int main (int argc, char *argv [])
{
    int i;                              /* Loop ounter */
    int ecount = 0;                     /* Error count */
    int wcount = 0;                     /* Warn count */

    uErrorSet (20, &ecount);            /* Max Num entries + counter */
    uWarnSet (&wcount);                 /* Warning count */
    uUtilitySet (progname);             /* Set name of program */
    uVerboseSet (1);                    /* Verbose setting */

    while (1) {
        i = getopt (argc, argv, "be:E:Ih?qv:o:");
        if (i == EOF)
            break;
        switch (i) {
        case 'b':
            useBase = 1;
            break;
        case 'e':
            uErrorChannelSet (optarg, UERROR_LOG_CREATE);
            break;
        case 'E':
            uErrorChannelSet (optarg, UERROR_LOG_APPEND);
            break;
        case 'h':
        case '?':
            usage();
            break;
        case 'I':
            fprintf (stdout, "%s. Version %s. %s.\n",
                     progname, MODULE_VERSION, __DATE__);
            exit (0);
            break;
        case 'o':
            outFile = optarg;
            break;
        case 'q':
            uInteractiveSet (0);        /* Disable interactive mode */
            break;
        case 'v':
            /* Verbose setting */
            uVerboseSet (optarg[0] - '0'); 
            break;
        default:
            uError ("Cannot understand option [-%c]\n", i);
            usage();
            break;
        }
    }

    uOpenErrorChannel ();
    argv = getfiles (&argc, argv, optind);

    if (outFile == NULL)
    {
        fo = stdout;
        uInteractiveSet (0);
    }
    else if ((fo = fopen (outFile, "wb")) == NULL)
        uFatal("Cannot open output file [%s]\n", outFile);

    /* Process all of the files */

    uVerbose (0, "Pass 1 - Collecting Symbols\n");
    for (i = 1; i < argc; i++)
        processFile (1, argv[i]);

    uVerbose (0, "Pass 2 - Generating files.\n");
    for (i = 1; i < argc; i++)
        processFile (2, argv[i]);
    if (fo != NULL)
    {
        fprintf (fo, "%s", eofStr);
        fclose (fo);
    }

    uCloseErrorChannel ();
    return (ecount);
}
